apiservice/apiservice.go

```diff
@@ -0,0 +1,275 @@
+package apiservice
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"google.golang.org/grpc/connectivity"
+
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/credentials"
+	"google.golang.org/grpc/status"
+
+	"github.com/onflow/flow/protobuf/go/flow/access"
+
+	"github.com/onflow/flow-go/engine/access/rpc/backend"
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/utils/grpcutils"
+)
+
+func NewFlowCachedAccessAPI(accessNodeAddressAndPort flow.IdentityList, timeout time.Duration) (*FlowCachedAccessAPI, error) {
+	ret := &FlowCachedAccessAPI{}
+	ret.timeout = timeout
+	ret.upstream = make([]access.AccessAPIClient, accessNodeAddressAndPort.Count())
+	ret.connections = make([]*grpc.ClientConn, accessNodeAddressAndPort.Count())
+	for i, identity := range accessNodeAddressAndPort {
+		// Store the client setup parameters such as address, public, key and timeout, so that
+		// we can refresh the API on connection loss
+		ret.ids[i] = identity
+
+		// We fail on any single error on startup, so that
+		// we identify bootstrapping errors early
+		err := ret.updateClient(i)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	ret.roundRobin = 0
+	return ret, nil
+}
+
+type FlowCachedAccessAPI struct {
```

Comment:
I think using `Cached` in the name is misleading since it doesn't include any caching. How about using `Proxy` or `Proxied` instead

Reply:
Addressed: https://github.com/GetElastech/flow-go/commit/c2cef1d25ef4951699ab09828175c9d01b3dc061